clang/tools/driver/CachedDiagnostics.cpp

```diff
@@ -0,0 +1,741 @@
+    if (io.outputting()) {
+      if (s.Buffer) {
+        std::string EncodedContents = encodeBase64(s.Buffer.value());
+        io.mapRequired("buffer", EncodedContents);
+      }
+    } else {
+      std::optional<std::string> EncodedContents;
+      io.mapOptional("buffer", EncodedContents);
+      if (EncodedContents) {
+        std::vector<char> Decoded;
+        cantFail(decodeBase64(*EncodedContents, Decoded));
+        s.Buffer = StringRef(Decoded.data(), Decoded.size());
+      }
+    }
+  }
+};
+
+template <> struct MappingTraits<cached_diagnostics::SLocEntry::ExpansionInfo> {
+  static void mapping(IO &io, cached_diagnostics::SLocEntry::ExpansionInfo &s) {
+    io.mapRequired("length", s.Length);
+    io.mapOptional("spelling_loc", s.SpellingLoc);
+    io.mapOptional("expansion_start_loc", s.ExpansionStartLoc);
+    io.mapOptional("expansion_end_loc", s.ExpansionEndLoc);
+    io.mapRequired("is_token_range", s.IsTokenRange);
+  }
+};
+
+template <> struct MappingTraits<cached_diagnostics::Location> {
+  static void mapping(IO &io, cached_diagnostics::Location &s) {
+    io.mapRequired("sloc_idx", s.SLocIdx);
+    io.mapRequired("offset", s.Offset);
+  }
+};
+
+template <> struct MappingTraits<cached_diagnostics::Range> {
+  static void mapping(IO &io, cached_diagnostics::Range &s) {
+    io.mapOptional("begin", s.Begin);
+    io.mapOptional("end", s.End);
+    io.mapRequired("is_token_range", s.IsTokenRange);
+  }
+};
+
+template <> struct MappingTraits<cached_diagnostics::FixItHint> {
+  static void mapping(IO &io, cached_diagnostics::FixItHint &s) {
+    io.mapOptional("remove_range", s.RemoveRange);
+    io.mapOptional("insert_range", s.InsertFromRange);
+    io.mapOptional("code", s.CodeToInsert);
+    io.mapRequired("before_previous", s.BeforePreviousInsertions);
+  }
+};
+
+template <> struct MappingTraits<cached_diagnostics::Diagnostic> {
+  static void mapping(IO &io, cached_diagnostics::Diagnostic &s) {
+    io.mapRequired("id", s.ID);
+    io.mapRequired("level", (unsigned &)s.Level);
+    io.mapRequired("message", s.Message);
+    io.mapOptional("loc", s.Loc);
+    io.mapOptional("ranges", s.Ranges);
+    io.mapOptional("fixits", s.FixIts);
+  }
+};
+
+template <> struct MappingTraits<cached_diagnostics::Diagnostics> {
+  static void mapping(IO &io, cached_diagnostics::Diagnostics &s) {
+    io.mapRequired("sloc_entries", s.SLocEntries);
+    io.mapRequired("diagnostics", s.Diags);
+  }
+};
+} // namespace llvm::yaml
+
+LLVM_YAML_IS_SEQUENCE_VECTOR(cached_diagnostics::SLocEntry)
+LLVM_YAML_IS_SEQUENCE_VECTOR(cached_diagnostics::Diagnostic)
+LLVM_YAML_IS_SEQUENCE_VECTOR(cached_diagnostics::Range)
+LLVM_YAML_IS_SEQUENCE_VECTOR(cached_diagnostics::FixItHint)
+
+std::optional<std::string>
+CachedDiagnosticSerializer::serializeEmittedDiagnostics() {
+  if (empty())
+    return std::nullopt;
+
+  SmallString<512> Buf;
+  raw_svector_ostream OS(Buf);
+  yaml::Output YOut(OS);
+  YOut << CachedDiags;
```

Comment:
This is more complex than just sorting a vector. The diagnostic "notes" are implicitly associated with the non-"note" diagnostic that preceded them, so you need to keep track of "notes" separately from the other diagnostics, then order the non-"notes" diagnostics only, then re-insert the associated "notes" after the right diagnostics in the same order.

I'd like to avoid it for now, but it's something to keep in mind in case it becomes difficult to keep diagnostic emission deterministic.